node/libs/concurrency/src/sync/prunable_queue.rs

```diff
@@ -0,0 +1,148 @@
+    has_items_send: watch::Sender<bool>,
+}
+
+pub struct Sender<T, U> {
+    shared: Arc<Shared<T, U>>,
+    pruning_predicate: Box<dyn Sync + Send + Fn(&T, &T) -> bool>,
+}
+
+impl<T, U> Sender<T, U> {
+    pub async fn enqueue(&self, item: T) -> oneshot::Receiver<U> {
+        // Create oneshot channel for returning result asynchronously.
+        let (res_send, res_recv) = oneshot::channel();
+
+        let mut queue = self.shared.queue.lock().await;
+        queue.retain(|existing_item| !(self.pruning_predicate)(&existing_item.0, &item));
+        queue.push_back((item, res_send));
+
+        // Ignore sending error.
+        let _ = self.shared.has_items_send.send(true);
+
+        res_recv
+    }
+}
+
+pub struct Receiver<T, U> {
+    shared: Arc<Shared<T, U>>,
+    has_items_recv: watch::Receiver<bool>,
+}
+
+impl<T, U> Receiver<T, U> {
+    pub async fn dequeue(&mut self, ctx: &Ctx) -> ctx::OrCanceled<(T, oneshot::Sender<U>)> {
+        sync::wait_for(ctx, &mut self.has_items_recv, |has_items| *has_items).await?;
+        let mut queue = self.shared.queue.lock().await;
+        // `None` is unexpected because we waited for new items.
+        let item = queue.pop_front().unwrap();
+
+        if queue.len() == 0 {
+            // Send error is unexpected because `self` holds the receiver.
+            self.shared.has_items_send.send(false).unwrap();
+        }
+
+        Ok(item)
+    }
+}
+
+// Test scenario:
+// Enqueue two sets of 0..1000 items, in conjunction, while pruning
+// so that only one 0..1000 set is expected to remain in queue.
+// Then, dequeue to assert the queue's content.
+#[tokio::test]
+async fn test_prunable_queue() {
+    use tokio::time::{timeout, Duration};
+
+    #[derive(Debug, Clone)]
+    struct ItemType(usize, usize);
+
+    let ctx = ctx::test_root(&ctx::RealClock);
+
+    let (sender, mut receiver) = new(Box::new(|a: &ItemType, b: &ItemType| a.0 != b.0));
+
+    let sender1 = Arc::new(sender);
+    let sender2 = sender1.clone();
+
+    let handle1 = tokio::spawn(async move {
+        let set = 1;
+        let items = (0..1000).map(|i| ItemType(i, set));
+        for item in items {
+            let _ = sender1.enqueue(item).await;
+            tokio::task::yield_now().await;
+        }
+    });
+    let handle2 = tokio::spawn(async move {
+        let set = 2;
+        let items = (0..1000).map(|i| ItemType(i, set));
+        for item in items {
+            let _ = sender2.enqueue(item).await;
+            tokio::task::yield_now().await;
+        }
+    });
+    tokio::try_join!(handle1, handle2).unwrap();
+
+    tokio::spawn(async move {
+        let mut i = 0;
+        loop {
+            let (item, sender) = receiver.dequeue(&ctx).await.unwrap();
+            assert_eq!(item.0, i);
+            let _ = sender.send(());
+
+            i = i + 1;
+            if i == 1000 {
+                match timeout(Duration::from_secs(0), receiver.dequeue(&ctx)).await {
```

Comment:
afair tokio::time::timeout doesn't specify the semantics when timeout is 0 AND future completes immediately. IMO it would be more beneficial to run receiver and sender (adding additional items) in parallel (on a prepopulated queue, which you already prepared) to make sure that consumer doesn't overtake the producer. Alternatively you can add a "try_recv" method for checking that the queue is actually empty.

Reply:
> afair tokio::time::timeout doesn't specify the semantics when timeout is 0 AND future completes immediately.

> Alternatively you can add a "try_recv" method for checking that the queue is actually empty.

I came this solution due to the no-guarantees in `tokio::select!` which were noticeable, while this solution seems stable. Looking at the code, it indeed tries first to poll the future, and only if not ready proceeds to the deadline. So it might be better than adding a "peek"-style method which isn't needed otherwise. 

> IMO it would be more beneficial to run receiver and sender (adding additional items) in parallel (on a prepopulated queue, which you already prepared) to make sure that consumer doesn't overtake the producer.

I agree that the test case should be improved. It was just for having something basic until getting a review ack.

Reply:
Test code changes here: https://github.com/matter-labs/era-consensus/pull/56/commits/020f25b2049f9f4bb2dbbddc9a0e876eadaf5389